add test test_execute_with_lamports_and_instruction_struct

IvanAdamovNeon opened the pull request.

IvanAdamovNeon committed:
add test test_execute_with_lamports_and_instruction_struct

romanova-natasha commented on contracts/precompiled/CallSolanaCaller.sol:
```diff
@@ -57,6 +58,20 @@ contract CallSolanaCaller {
+        );
+        emit LogBytes(returnData);
+    }
+
```
There is another method `executeWithSeed` with ICallSolana.Instruction arg (in the contract ICallSolana.sol)
I think it would be better to also cover it with a separate test (and we can add another instruction: COUNTER or transfer).

Commit:
fix serialized tx response, minor refactoring, add execute_with_seed method check

fix serialized tx response, minor refactoring, add execute_with_seed method check

Commit:
do not serialize data for transfer instruction

do not serialize data for transfer instruction

Commit:
Merge remote-tracking branch 'origin/develop' into NDEV-3824-composability-tests

# Conflicts:
#	integration/tests/neon_evm/test_interoperability.py
#	integration/tests/neon_evm/utils/ethereum.py

Commit:
fix after merge develop

fix after merge develop

kristinaNikolaevaa commented on utils/helpers.py:
```diff
@@ -258,6 +258,17 @@ def serialize_instruction(program_id: Pubkey, instruction) -> bytes:
+    for key in instruction["accounts"]:
+        serialized_accounts.append((solana_pubkey_to_bytes32(key.pubkey), key.is_signer, key.is_writable))
+
+    serialized_data = len(instruction["instruction_data"]).to_bytes(8, "little") + instruction["instruction_data"]
```
Is it correct? adding len of instruction to data?

romanova-natasha replied:
Fixed
I just decided that it won't work in case of other instructions (not transfer, where I added additional check). But I was wrong, we do not need to add len of the data here.

kristinaNikolaevaa replied:
strange that other added tests passed

romanova-natasha committed:
do not serialize data for instruction type arg

do not serialize data for instruction type arg

kristinaNikolaevaa commented on integration/tests/neon_evm/test_interoperability.py:
```diff
@@ -111,6 +111,79 @@ def test_execute_from_instruction_for_compute_budget(
+        solana_caller,
+        solana_client,
+    ):
+        instruction_struct = {
```
for me it's a bit strange decision to create a custom structure like that and then use decirialization to prepare it for function params serialize_instruction_struct()
More logical to use inner python solana lib class Instruction which we use everywhere 

Also I don't like duplications in execute_with_seed_and_instruction_struct and execute_with_instruction_struct() 
There are just big part of code are the same ias in next exist functions

And  tests itself, old tests could be just parametrised 

But these notes are not critical

romanova-natasha replied:
I replaced custom structure with solana lib `Instruction(...)`. 

We need `serialize_instruction_struct()` method to prepare arg with struct type `ICallSolana.Instruction`, cause we need to pass a tuple (serialized_program_is, serialized_accounts, serialized_data) as the struct. 

We have `execute`, `execute_with_seed`, `execute_with_instruction_struct`, `execute_with_seed_and_instruction_struct` methods. There are duplications in these methods, but they are also parametrized with `lamports` and `is_iterative` options, it already makes them a bit overloaded with logic. I'll add a note about this to the task with refactoring and try to figure out how to improve this part of code.

romanova-natasha committed:
rm custom instruction struct

rm custom instruction struct

romanova-natasha committed:
Merge branch 'develop' into NDEV-3824-composability-tests
